sphinx/builders/linkcheck.py

```diff
@@ -473,7 +473,7 @@ def _check_uri(self, uri: str, hyperlink: Hyperlink) -> tuple[str, str, int]:
                 ) as response:
                     if (self.check_anchors and response.ok and anchor
                             and not contains_anchor(response, anchor)):
-                        raise Exception(__(f'Anchor {quote(anchor)!r} not found'))
+                        raise Exception(__("Anchor '%s' not found") % quote(anchor))
```

Comment:
Would using `%r` be closer to the semantic equivalent?
```suggestion
                        raise Exception(__('Anchor %r not found') % quote(anchor))
```

Reply:
Hrm - perhaps, although it seems they're slightly different operations?  Using [`!r` in an f-string](https://docs.python.org/3.12/reference/lexical_analysis.html#f-strings) calls `repr(...)` -- without specifying a string encoding, in contrast to [`%r` in printf style](https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting) that _additionally_ encodes as `ascii`.

The [`urllib.parse.quote`](https://docs.python.org/3.12/library/urllib.parse.html#urllib.parse.quote) call should handle URL-encoding the anchor name for us, so I don't think it's necessary to repeat-encode into ASCII.  But I'm still thinking it through..

Reply:
There's little harm or cost in repeating ASCII-encoding of the result I suppose, and it makes the diff simpler and clearer.  The cost is tiny compared to any network requests and HTML parsing.  So: OK, I think this makes sense.

Reply:
Note also: need to be a little careful to retain the correct quoting _within_ the string literal so that the existing translation resources match.

Reply:
Ah, there's a problem: we want to keep the single-quotes in the original, pre-localization string literal.  Using `%r` does add single-quotes as a side-effect of calling `repr(...)` on a `str` value, but we want those single-quotes in place before localization takes place.

Reply:
@AA-Turner I'm relatively confident that we need the single-quotes in the string literal, and so using `'%s'` (as-is in the branch currently) is the way to go.

Reply:
> _additionally_ encodes as `ascii`

Isn't that ``%a``?

```pycon
>>> for conv in ('%s', '%r', '%a'): print(f'{conv}: ' + conv % "ü")
%s: ü
%r: 'ü'
%a: '\xfc'
```

Reply:
Yep, that's correct - my mistake.

Even so, I think it is `%s` that we want (no `repr` call) in the literal.